src/ProducerEventProcessor/ProducerEventProcessorTest.cpp

```diff
@@ -1321,33 +1324,189 @@ TEST(ProducerEventProcessor, ThreadNameSmoke) {
+
+  ProducerCaptureEvent thread_state_slice_callstack_producer_event1;
+  ThreadStateSliceCallstack* thread_state_slice_callstack1 =
+      thread_state_slice_callstack_producer_event1.mutable_thread_state_slice_callstack();
+  thread_state_slice_callstack1->set_thread_state_slice_tid(kTid1);
+  thread_state_slice_callstack1->set_timestamp_ns(kTimestampNs1 - kDurationNs1);
+  Callstack* callstack1 = thread_state_slice_callstack1->mutable_callstack();
+  callstack1->add_pcs(kCallstackFrame1);
+  callstack1->add_pcs(kCallstackFrame2);
+  callstack1->add_pcs(kCallstackFrame3);
+
+  ProducerCaptureEvent thread_state_slice_producer_event1;
+  ThreadStateSlice* thread_state_slice1 =
+      thread_state_slice_producer_event1.mutable_thread_state_slice();
+  thread_state_slice1->set_pid(kPid1);
+  thread_state_slice1->set_tid(kTid1);
+  thread_state_slice1->set_thread_state(ThreadStateSlice::kRunnable);
+  thread_state_slice1->set_duration_ns(kDurationNs1);
+  thread_state_slice1->set_end_timestamp_ns(kTimestampNs1);
+  thread_state_slice1->set_switch_out_or_wakeup_callstack_status(
+      ThreadStateSlice::kWaitingForCallstack);
+  thread_state_slice1->set_switch_out_or_wakeup_callstack_id(0);
+  ThreadStateSlice expected_thread_state_slice1 = *thread_state_slice1;
+
+  ProducerCaptureEvent thread_state_slice_callstack_producer_event2;
+  ThreadStateSliceCallstack* thread_state_slice_callstack2 =
+      thread_state_slice_callstack_producer_event2.mutable_thread_state_slice_callstack();
+  thread_state_slice_callstack2->set_thread_state_slice_tid(kTid2);
+  thread_state_slice_callstack2->set_timestamp_ns(kTimestampNs2 - kDurationNs2);
+  Callstack* callstack2 = thread_state_slice_callstack2->mutable_callstack();
+  callstack2->add_pcs(kCallstackFrame1);
+  callstack2->add_pcs(kCallstackFrame2);
+  callstack2->add_pcs(kCallstackFrame3);
+
+  ProducerCaptureEvent thread_state_slice_producer_event2;
+  ThreadStateSlice* thread_state_slice2 =
+      thread_state_slice_producer_event2.mutable_thread_state_slice();
+  thread_state_slice2->set_pid(kPid1);
+  thread_state_slice2->set_tid(kTid2);
+  thread_state_slice2->set_thread_state(ThreadStateSlice::kRunnable);
+  thread_state_slice2->set_duration_ns(kDurationNs2);
+  thread_state_slice2->set_end_timestamp_ns(kTimestampNs2);
+  thread_state_slice2->set_switch_out_or_wakeup_callstack_status(
+      ThreadStateSlice::kWaitingForCallstack);
+  thread_state_slice2->set_switch_out_or_wakeup_callstack_id(0);
+  ThreadStateSlice expected_thread_state_slice2 = *thread_state_slice2;
+
+  std::vector<ClientCaptureEvent> actual_client_capture_events;
+
+  uint64_t actual_callstack_key = 0;
+  EXPECT_CALL(collector, AddEvent)
+      .Times(3)
+      .WillRepeatedly(Invoke([&actual_client_capture_events,
+                              &actual_callstack_key](ClientCaptureEvent&& client_capture_event) {
+        if (client_capture_event.has_interned_callstack()) {
+          ASSERT_EQ(actual_callstack_key, 0);
```

Comment:
Could you add a comment that you only expect one interned callstack? I was confused why this would hold the second time you come here.

Reply:
Done.